internal/exec/stages/files/units.go

```diff
@@ -17,57 +17,145 @@ package files
+		systemdVersion = version[1 : len(version)-1]
+	}
+	if value, err := strconv.Atoi(systemdVersion); err == nil {
+		if value < 240 {
+			if err := s.Logger.Warning("The version of systemd is less than 240, Enabling/disabling instantiated units may not work. See https://github.com/coreos/ignition/issues/586 for more information."); err != nil {
+				return err
+			}
+		}
+	} else {
+		return err
+	}
+	return nil
+}
+
 // createUnits creates the units listed under systemd.units.
 func (s *stage) createUnits(config types.Config) error {
-	enabledOneUnit := false
+	presets := make(map[string]*Preset)
 	for _, unit := range config.Systemd.Units {
 		if err := s.writeSystemdUnit(unit, false); err != nil {
 			return err
 		}
 		if unit.Enabled != nil {
 			if *unit.Enabled {
-				if err := s.Logger.LogOp(
-					func() error { return s.EnableUnit(unit) },
-					"enabling unit %q", unit.Name,
-				); err != nil {
-					return err
+				if strings.Contains(unit.Name, "@") {
+					at := strings.Index(unit.Name, "@")
+					dot := strings.Index(unit.Name, ".")
+					instance := unit.Name[at+1 : dot]
+					serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
+					if _, ok := presets[serviceInstance]; ok {
+						presets[serviceInstance].instances = append(presets[serviceInstance].instances, instance)
+					} else {
+						presets[serviceInstance] = &Preset{serviceInstance, true, true, []string{instance}}
+					}
+				} else {
+					presets[unit.Name] = &Preset{unit.Name, true, false, []string{}}
 				}
+
 			} else {
+				if strings.Contains(unit.Name, "@") {
+					at := strings.Index(unit.Name, "@")
+					dot := strings.Index(unit.Name, ".")
+					instance := unit.Name[at+1 : dot]
+					serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
+					presets[serviceInstance] = &Preset{serviceInstance, false, true, []string{instance}}
+				} else {
+					presets[unit.Name] = &Preset{unit.Name, false, false, []string{}}
+				}
+			}
+			if unit.Mask != nil && *unit.Mask {
+				relabelpath := ""
 				if err := s.Logger.LogOp(
-					func() error { return s.DisableUnit(unit) },
-					"disabling unit %q", unit.Name,
+					func() error {
+						var err error
+						relabelpath, err = s.MaskUnit(unit)
+						return err
+					},
+					"masking unit %q", unit.Name,
 				); err != nil {
 					return err
 				}
+				s.relabel(relabelpath)
 			}
-			enabledOneUnit = true
 		}
-		if unit.Mask != nil && *unit.Mask {
-			relabelpath := ""
+	}
+	if err := s.createSystemdPresetFile(presets); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createSystemdPresetFile creates the presetfile for enabled/disabled
+// instantiated/systemd units.
+func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
+	var instantiateEnableUnit, instantiateDisableUnit string
+	for _, value := range presets {
+		if value.enabled {
+			if value.instantiateable {
+				s.isRequiredSystemdVersionExists()
```

Comment:
let's try to only print out this warning once.

Reply:
Fixed.